File: backend/app/schemas/mission.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, computed_field

from app.core.mission_engine import is_same_period


class CreateMissionRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    description: str | None = Field(None, max_length=2000)
    category: Literal[
        "vitality", "strength", "intelligence", "wisdom", "charisma", "discipline"
    ]
    difficulty: Literal["easy", "medium", "hard", "epic"]
    frequency: Literal["daily", "weekly", "monthly"]
    target_count: int = Field(0, ge=0, description="0 = unlimited")


class UpdateMissionRequest(BaseModel):
    title: str | None = Field(None, min_length=1, max_length=200)
    description: str | None = Field(None, max_length=2000)
    target_count: int | None = Field(None, ge=0)
    # "completed" is auto-set when target_count is reached; not user-settable via PATCH.
    status: Literal["active", "paused", "archived"] | None = None


class CheckinRequest(BaseModel):
    evidence_text: str = Field(
        ...,
        min_length=20,
        max_length=5000,
        description="What you actually did — minimum 20 characters.",
    )
    reflection: str | None = Field(None, max_length=5000)
    effort_level: int = Field(..., ge=1, le=5)


class MissionLogResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    mission_id: uuid.UUID
    evidence_text: str
    reflection: str | None
    effort_level: int
    xp_awarded: int
    streak_at_checkin: int
    created_at: datetime


class MissionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    title: str
    description: str | None
    category: str
    difficulty: str
    frequency: str
    status: str
    target_count: int
    completion_count: int
    current_streak: int
    longest_streak: int
    last_completed_at: datetime | None
    created_at: datetime
    updated_at: datetime

    @computed_field
    @property
    def can_checkin_now(self) -> bool:
        """Whether the user is eligible to check in right now."""
        if self.status != "active":
            return False
        if self.last_completed_at is None:
            return True
        return not is_same_period(self.last_completed_at, self.frequency)


class CheckinResponse(BaseModel):
    mission: MissionResponse
    log: MissionLogResponse
    xp_awarded: int
    new_streak: int
    levels_gained: list[int]


class MissionListResponse(BaseModel):
    missions: list[MissionResponse]
    total: int
    page: int
    page_size: int


class MissionLogListResponse(BaseModel):
    logs: list[MissionLogResponse]
    total: int
    page: int
    page_size: int

File: mobile/lib/core/models/mission.dart
class MissionResponse {
  final String id;
  final String userId;
  final String title;
  final String? description;
  final String category;
  final String frequency;
  final String difficulty;
  final String status;
  final int targetCount;
  final int completionCount;
  final int currentStreak;
  final int longestStreak;
  final DateTime? lastCompletedAt;
  final bool canCheckinNow;
  final DateTime createdAt;
  final DateTime updatedAt;

  const MissionResponse({
    required this.id,
    required this.userId,
    required this.title,
    this.description,
    required this.category,
    required this.frequency,
    required this.difficulty,
    required this.status,
    required this.targetCount,
    required this.completionCount,
    required this.currentStreak,
    required this.longestStreak,
    this.lastCompletedAt,
    required this.canCheckinNow,
    required this.createdAt,
    required this.updatedAt,
  });

  factory MissionResponse.fromJson(Map<String, dynamic> json) => MissionResponse(
        id:               json['id'] as String,
        userId:           json['user_id'] as String,
        title:            json['title'] as String,
        description:      json['description'] as String?,
        category:         json['category'] as String,
        frequency:        json['frequency'] as String,
        difficulty:       json['difficulty'] as String,
        status:           json['status'] as String,
        targetCount:      json['target_count'] as int,
        completionCount:  json['completion_count'] as int,
        currentStreak:    json['current_streak'] as int,
        longestStreak:    json['longest_streak'] as int,
        lastCompletedAt:  json['last_completed_at'] != null
            ? DateTime.parse(json['last_completed_at'] as String)
            : null,
        canCheckinNow:    (json['can_checkin_now'] as bool?) ?? true,
        createdAt:        DateTime.parse(json['created_at'] as String),
        updatedAt:        DateTime.parse(json['updated_at'] as String),
      );

  bool get isActive => status == 'active';
}

class MissionTemplate {
  final String id;
  final String title;
  final String description;
  final String category;
  final String difficulty;
  final String frequency;
  final String emoji;
  final int sortOrder;

  const MissionTemplate({
    required this.id,
    required this.title,
    required this.description,
    required this.category,
    required this.difficulty,
    required this.frequency,
    required this.emoji,
    required this.sortOrder,
  });

  factory MissionTemplate.fromJson(Map<String, dynamic> json) => MissionTemplate(
        id:          json['id'] as String,
        title:       json['title'] as String,
        description: json['description'] as String,
        category:    json['category'] as String,
        difficulty:  json['difficulty'] as String,
        frequency:   json['frequency'] as String,
        emoji:       json['emoji'] as String,
        sortOrder:   json['sort_order'] as int,
      );
}

class CreateMissionRequest {
  final String title;
  final String? description;
  final String category;
  final String frequency;
  final String difficulty;

  const CreateMissionRequest({
    required this.title,
    this.description,
    required this.category,
    required this.frequency,
    required this.difficulty,
  });

  Map<String, dynamic> toJson() => {
        'title': title,
        if (description?.isNotEmpty == true) 'description': description,
        'category': category,
        'frequency': frequency,
        'difficulty': difficulty,
        'target_count': 0,
      };
}

class CheckinRequest {
  final String evidenceText;
  final String? reflection;
  final int effortLevel;

  const CheckinRequest({
    required this.evidenceText,
    this.reflection,
    required this.effortLevel,
  });

  Map<String, dynamic> toJson() => {
        'evidence_text': evidenceText,
        if (reflection?.isNotEmpty == true) 'reflection': reflection,
        'effort_level': effortLevel,
      };
}

class CheckinResponse {
  final MissionResponse mission;
  final int xpAwarded;
  final int newStreak;
  final List<int> levelsGained;

  const CheckinResponse({
    required this.mission,
    required this.xpAwarded,
    required this.newStreak,
    required this.levelsGained,
  });

  factory CheckinResponse.fromJson(Map<String, dynamic> json) => CheckinResponse(
        mission:     MissionResponse.fromJson(json['mission'] as Map<String, dynamic>),
        xpAwarded:   json['xp_awarded'] as int,
        newStreak:   json['new_streak'] as int,
        levelsGained:(json['levels_gained'] as List).cast<int>(),
      );

  bool get leveledUp => levelsGained.isNotEmpty;
}
